### PR TITLE
Migrate to GitHub Actions from CircleCI

### LINKED ISSUE #4137: Help us migrate to GitHub Actions
## Description

We're looking for community support to help us migrate from CircleCI to GitHub Actions.

The CircleCI file is here:
https://github.com/facebook/react-native-website/blob/main/.circleci/config.yml

It's pretty simple so it should be easy to migrate over to GitHub Actions.

If you intend to work on it please comment before starting the work.

### PR DESCRIPTION
**Pull request by apicgg**
Fixes #4137 

- [x] Set up Node environment.
- [x] Run lint script.
- [x] Run build script.

### CODE REVIEW

File: .github/workflows/cicd.yml [hunk 1]
```diff
@@ -0,0 +1,69 @@
+##CICD Version: v0.0.1
+
+name: GitHub_CICD
```
Comment (cortinico) [inline .github/workflows/cicd.yml:3]:
```suggestion
name: Pre-merge
```

File: .github/workflows/cicd.yml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        uses: actions/setup-node@v4
+
+      - name: Checkout repository
+        uses: actions/checkout@v4
+
+      - name: Save repository to cache
+        uses: actions/cache@v4
+        with:
+          path: ~/react-native-website
+          key: v1-repo-${{ github.sha }}
```
Comment (cortinico) [inline .github/workflows/cicd.yml:24]: Let's remove all the `actions/cache` ok?

File: .github/workflows/cicd.yml [hunk 3]
```diff
@@ -0,0 +1,69 @@
+##CICD Version: v0.0.1
```
Comment (cortinico) [inline .github/workflows/cicd.yml:1]:
```suggestion
```

**Review by cortinico — COMMENTED** [review]
Also can we call the file `pre-merge.yml`?

**Review by cortinico — COMMENTED** [review]
Also can we call the file `pre-merge.yml`?

**Comment by apicgg** [comment]
@cortinico Done as per your comments.

**Comment by cortinico** [comment]
@apicgg seems like both `build` and `lint` are failing. Can you look into them and make them green?

**Comment by apicgg** [comment]
> @apicgg seems like both `build` and `lint` are failing. Can you look into them and make them green?

@cortinico Hey, yeah seems like the GitHub runner is picking up wrong path for the repo. I am checking this right now on how to fix.

**Comment by Simek** [comment]
> I am checking this right now on how to fix.

We can use the [`working-directory: docs`](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsworking-directory) param for the scripts that need to be run from `docs` directory.

**Comment by apicgg** [comment]
> > I am checking this right now on how to fix.
> 
> We can use the [`working-directory: docs`](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsworking-directory) param for the scripts that need to be run from `docs` directory.

@Simek Any idea why GH Actions is appending the working directory like below? For reference - [Action Link](https://github.com/facebook/react-native-website/actions/runs/9469602554/job/26088678319)

image

**Comment by Simek** [comment]
This is probably throwing due to the configuration mistake. I would recommend making the lint and build a separate steps in the same workflow not separated jobs. If you want to retain the current division, you need to move the setup step in into both of the jobs (`lint` and `build`). `needs` only expects the successful run on mentioned job, but the checkouts are not shared.

**Comment by apicgg** [comment]
@cortinico This is ready to review. The lint job is apparently taking much longer with default GitHub Actions runners.

File: .github/workflows/pre-merge.yml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    runs-on: ubuntu-latest
+    steps:
+      - name: Set up Node.js
+        uses: actions/setup-node@v4
+
+      - name: Checkout repository
+        uses: actions/checkout@v4
+
+      - name: Install dependencies
+        run: yarn install --frozen-lockfile --no-progress --non-interactive
```
Comment (Simek) [inline .github/workflows/pre-merge.yml:19]:
```suggestion
  setup:
    runs-on: ubuntu-latest
    steps:
      - name: Set up Node.js
        uses: actions/setup-node@v4

      - name: Checkout repository
        uses: actions/checkout@v4

      - name: Install dependencies
        run: yarn install --frozen-lockfile --no-progress --non-interactive
```

After inlining, we can remove that job.

File: .github/workflows/pre-merge.yml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+  lint:
+    runs-on: ubuntu-latest
+    needs: setup
```
Comment (Simek) [inline .github/workflows/pre-merge.yml:23]:
```suggestion
```

File: .github/workflows/pre-merge.yml [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+      - name: Install dependencies
+        run: yarn install --frozen-lockfile --no-progress --non-interactive
+
+      - name: Run Lint
+        run: yarn ci:lint
+        working-directory: website
+
+  build:
+    runs-on: ubuntu-latest
+    needs: setup
```
Comment (Simek) [inline .github/workflows/pre-merge.yml:40]:
```suggestion
```

**Comment by Simek** [comment]
@cortinico Do you also have in mind moving the deployments previews from CircleCI, or it was about PR checks only?

**Comment by apicgg** [comment]
@Simek I have pushed the changes now.

**Comment by cortinico** [comment]
> Do you also have in mind moving the deployments previews from CircleCI, or it was about PR checks only?

I don't think we do deployments of Previews from CircleCI.

**Comment by cortinico** [comment]
Thank you very much for your help @apicgg 🙏

**Pull request merged**